coil/src/main/java/dev/chrisbanes/accompanist/coil/Coil.kt

```diff
@@ -186,36 +188,50 @@ fun CoilImage(
         else -> null
     }
 
-    // Update the modifier, so that we listen for size changes to invoke the actor
-    val mod = modifier.onSizeChanged { size ->
+    WithConstraints(modifier) {
+        val lastRequestedSize = remember(requestActor) { MutableRef(IntSize.Zero) }
+
+        val requestSize = IntSize(
+            width = if (constraints.hasBoundedWidth) constraints.maxWidth else UNSPECIFIED,
+            height = if (constraints.hasBoundedHeight) constraints.maxHeight else UNSPECIFIED
+        )
+
         val r = result
-        if (r == null || shouldRefetchOnSizeChange(r, size)) {
-            // If we don't have a result yet, or the size has changed and we should re-fetch,
-            // send it to the request actor
-            requestActor.send(size)
+        if (lastRequestedSize.value != requestSize &&
+            (r == null || shouldRefetchOnSizeChange(r, requestSize))
+        ) {
+            requestActor.send(requestSize)
+            lastRequestedSize.value = requestSize
         }
-    }
 
-    if (painter == null) {
-        // If we don't have a result painter, we add a Box with our modifier
-        Box(mod) {
-            // If we don't have a result yet, we can show the loading content
-            // (if not null)
-            if (result == null && loading != null) {
-                loading()
+        if (painter == null) {
+            // If we don't have a result painter, we add a Box with our modifier
+            Box {
+                // If we don't have a result yet, we can show the loading content
+                // (if not null)
+                if (result == null && loading != null) {
+                    loading()
+                }
             }
+        } else {
+            Image(
+                painter = painter,
+                contentScale = contentScale,
+                alignment = alignment,
+                colorFilter = colorFilter,
+            )
         }
-    } else {
-        Image(
-            painter = painter,
-            contentScale = contentScale,
-            alignment = alignment,
-            colorFilter = colorFilter,
-            modifier = mod
-        )
     }
 }
 
+/**
+ * Value for a [IntSize] dimension, where the dimension is not specified or is unknown.
+ */
+private const val UNSPECIFIED = -1
+
+@Stable
+private data class MutableRef<T>(var value: T)
```

Comment:
Smart! Nit: Add some comments about why you need this?

Reply:
Thanks, done in 05f1927